From 65867e8285e21de2fa793495beed16e4e97e109f Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 18 Sep 2014 14:20:16 -0400
Subject: [fs] xprtrdma: Handle additional connection events

Message-id: <1411050057-12417-54-git-send-email-steved@redhat.com>
Patchwork-id: 96218
O-Subject: [RHEL7.1 PATCH 53/94] xprtrdma: Handle additional connection events
Bugzilla: 1111708
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: Chuck Lever <chuck.lever@oracle.com>

Commit 38ca83a5 added RDMA_CM_EVENT_TIMEWAIT_EXIT. But that status
is relevant only for consumers that re-use their QPs on new
connections. xprtrdma creates a fresh QP on reconnection, so that
event should be explicitly ignored.

Squelch the alarming "unexpected CM event" message.

Signed-off-by: Chuck Lever <chuck.lever@oracle.com>
Tested-by: Steve Wise <swise@opengridcomputing.com>
Tested-by: Shirley Ma <shirley.ma@oracle.com>
Tested-by: Devesh Sharma <devesh.sharma@emulex.com>
Signed-off-by: Anna Schumaker <Anna.Schumaker@Netapp.com>
Upstream Id: 8079fb785e34de6dff34bd846b8b79c212861edf
Bugzilla: 1111708
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sunrpc/xprtrdma/verbs.c b/net/sunrpc/xprtrdma/verbs.c
index c2253d4..61c4129 100644
--- a/net/sunrpc/xprtrdma/verbs.c
+++ b/net/sunrpc/xprtrdma/verbs.c
@@ -326,8 +326,16 @@ static const char * const conn[] = {
  "rejected",
  "established",
  "disconnected",
- "device removal"
+ "device removal",
+ "multicast join",
+ "multicast error",
+ "address change",
+ "timewait exit",
 };
+
+#define CONNECTION_MSG(status)      \
+ ((status) < ARRAY_SIZE(conn) ?     \
+  conn[(status)] : "unrecognized connection error")
 #endif
 
 static int
@@ -385,23 +393,18 @@ rpcrdma_conn_upcall(struct rdma_cm_id *id, struct rdma_cm_event *event)
  case RDMA_CM_EVENT_DEVICE_REMOVAL:
   connstate = -ENODEV;
 connected:
-  dprintk("RPC:       %s: %s: %pI4:%u (ep 0x%p event 0x%x)\n",
-   __func__,
-   (event->event <= 11) ? conn[event->event] :
-      "unknown connection error",
-   &addr->sin_addr.s_addr,
-   ntohs(addr->sin_port),
-   ep, event->event);
   atomic_set(&rpcx_to_rdmax(ep->rep_xprt)->rx_buf.rb_credits, 1);
   dprintk("RPC:       %s: %sconnected\n",
      __func__, connstate > 0 ? "" : "dis");
   ep->rep_connected = connstate;
   ep->rep_func(ep);
   wake_up_all(&ep->rep_connect_wait);
-  break;
+  /*FALLTHROUGH*/
  default:
-  dprintk("RPC:       %s: unexpected CM event %d\n",
-   __func__, event->event);
+  dprintk("RPC:       %s: %pI4:%u (ep 0x%p): %s\n",
+   __func__, &addr->sin_addr.s_addr,
+   ntohs(addr->sin_port), ep,
+   CONNECTION_MSG(event->event));
   break;
  }
 
-- 
1.7.1